Add error handling for ALREADY_EXISTS in IAM CreateServiceAccount call

benhxy opened the pull request:
Introduce an optional resource argument to Google Cloud IAM Service Account. If `ignore_create_already_exists` is set to true, resource creation would succeed if response error is 409 ALREADY_EXISTS.

Fixes https://github.com/hashicorp/terraform-provider-google/issues/16185
Fixes https://github.com/hashicorp/terraform-provider-google/issues/10193

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:note
iam: introducde an optional resource argument to Google Cloud IAM Service Account. If `ignore_create_already_exists` is set to true, resource creation would succeed if response error is 409 ALREADY_EXISTS.
```

modular-magician commented:
Hello! I am a robot. It looks like you are a: ~Community Contributor~ Googler ~Core Contributor~. Tests will run automatically.

@rileykarson, a repository maintainer, has been assigned to review your changes. If you have not received review feedback within 2 business days, please leave a comment on this PR asking them to take a look.

You can help make sure that review is quick by [doing a self-review](https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/) and by [running impacted tests locally](https://googlecloudplatform.github.io/magic-modules/get-started/run-provider-tests/).

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 2 files changed, 33 insertions(+), 1 deletion(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 2 files changed, 33 insertions(+), 1 deletion(-))

benhxy commented:
To reviewer: an alternative argument name could be `create_ignore_already_exists`. Please advise if there are better names.

modular-magician commented:
#### Tests analytics 
Total tests: `135` 
Passed tests `103` 
Skipped tests: `30` 
Affected tests: `2` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

#### Action taken 
<details> <summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccServiceAccount_basic|TestAccDataSourceGoogleServiceAccountIdToken_impersonation </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccDataSourceGoogleServiceAccountIdToken_impersonation`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/d78223cb-1200-4a24-aa3b-47189be77bf3/recording/TestAccDataSourceGoogleServiceAccountIdToken_impersonation.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$ 
 `TestAccServiceAccount_basic`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/d78223cb-1200-4a24-aa3b-47189be77bf3/build-log/recording_build/TestAccServiceAccount_basic_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/d78223cb-1200-4a24-aa3b-47189be77bf3/recording/TestAccServiceAccount_basic.log)] 

$\textcolor{red}{\textsf{Please fix these to complete your PR.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/d78223cb-1200-4a24-aa3b-47189be77bf3/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/d78223cb-1200-4a24-aa3b-47189be77bf3/recording) for each test

rileykarson commented on mmv1/third_party/terraform/services/resourcemanager/resource_google_service_account_test.go:
```diff
@@ -88,6 +88,48 @@ func TestAccServiceAccount_basic(t *testing.T) {
+			},
+			// The second step creates a new resource that duplicates with the existing service account.
+			{
+				Config: testAccServiceAccountCreateIgnoreAlreadyExists(accountId, displayName, desc),
```
Oh- this will need to have both blocks! It will technically work almost all of the time as-is because the create and delete will happen simultaneously, but we can eliminate the flake by adding `resource "google_service_account" "acceptance"` in the second config too.

(deletes may cause errors though as we'll then try to delete the resource twice- I can't think of a great way to work around that w/o adding new user-facing functionality to the resource)

benhxy replied:
I see, thank you for the pointer.

I added both resources in the config block, and added a NOT_FOUND handling for DeleteServiceAccount.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 2 files changed, 65 insertions(+), 1 deletion(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 2 files changed, 65 insertions(+), 1 deletion(-))

modular-magician commented:
#### Tests analytics 
Total tests: `0` 
Passed tests `0` 
Skipped tests: `0` 
Affected tests: `0` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

$\textcolor{red}{\textsf{Errors occurred during REPLAYING mode. Please fix them to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/57563450-80c2-4210-ba4f-2c72b3d2f5a2/build-log/replaying_test.log)

benhxy commented on mmv1/third_party/terraform/services/resourcemanager/resource_google_service_account.go:
```diff
@@ -82,6 +83,12 @@ func ResourceGoogleServiceAccount() *schema.Resource {
 				Computed:    true,
 				Description: `The Identity of the service account in the form 'serviceAccount:{email}'. This value is often used to refer to the service account in order to grant IAM permissions.`,
 			},
+			"ignore_create_already_exists": {
```
Could you point me to where the markdown doc is?

rileykarson replied:
https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/third_party/terraform/website/docs/r/google_service_account.html.markdown

benhxy replied:
TY! Done

benhxy commented on mmv1/third_party/terraform/services/resourcemanager/resource_google_service_account_test.go:
```diff
@@ -166,6 +176,17 @@ resource "google_service_account" "acceptance" {
 
+func testAccServiceAccountIgnoreCreateAlreadyExists(account, name, desc string) string {
+	return fmt.Sprintf(`
+resource "google_service_account" "acceptance" {
```
Done

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 2 files changed, 75 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 2 files changed, 75 insertions(+), 2 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `0` 
Passed tests `0` 
Skipped tests: `0` 
Affected tests: `0` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

$\textcolor{red}{\textsf{Errors occurred during REPLAYING mode. Please fix them to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/848c3808-b208-4a33-a019-d5214831eae4/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 77 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 77 insertions(+), 2 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `0` 
Passed tests `0` 
Skipped tests: `0` 
Affected tests: `0` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

$\textcolor{red}{\textsf{Errors occurred during REPLAYING mode. Please fix them to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/adf6ebcc-25e0-471e-b88e-1c219baa7fdd/build-log/replaying_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 77 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 77 insertions(+), 2 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `0` 
Passed tests `0` 
Skipped tests: `0` 
Affected tests: `0` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

$\textcolor{red}{\textsf{Errors occurred during REPLAYING mode. Please fix them to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/85ac38de-beb5-4be2-a687-c2963e00b1fb/build-log/replaying_test.log)

benhxy commented:
Hmm I'm a bit stuck on the build error. When I clicked in the build log, it only has one line:
```
FAIL	github.com/hashicorp/terraform-provider-google-beta/google-beta/services/resourcemanager [build failed]
```

Locally, I followed https://googlecloudplatform.github.io/magic-modules/get-started/generate-providers/ to set up the environment.

Then with my change in magic-modules, I ran `make provider VERSION=beta OUTPUT_PATH="$GOPATH/src/github.com/hashicorp/terraform-provider-google-beta" PRODUCT=resourcemanager RESOURCE=google_service_account`. 

It showed a lot of `/Users/bensea/.rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/bundler-2.4.6/lib/bundler/shared_helpers.rb:58:in `getwd': Too many open files - getcwd (Errno::EMFILE)` but that probably didn't matter.

It also updated many other products even when I specified PRODUCT and RESOURCE filter.

But most importantly, it didn't update the files I edited in magic-modules:
```
terraform-provider-google-beta bensea$ git status | grep resourcemanager
	modified:   google-beta/services/resourcemanager/data_source_google_folders.go
	modified:   google-beta/services/resourcemanager/data_source_google_iam_testable_permissions_test.go
	modified:   google-beta/services/resourcemanager/data_source_google_organization.go
	modified:   google-beta/services/resourcemanager/data_source_google_organization_test.go
	modified:   google-beta/services/resourcemanager/data_source_google_project_organization_policy.go
	modified:   google-beta/services/resourcemanager/data_source_google_project_service.go
	modified:   google-beta/services/resourcemanager/data_source_google_project_service_test.go
	modified:   google-beta/services/resourcemanager/data_source_google_projects.go
	modified:   google-beta/services/resourcemanager/data_source_google_projects_test.go
	modified:   google-beta/services/resourcemanager/data_source_google_service_account.go
	modified:   google-beta/services/resourcemanager/data_source_google_service_account_access_token.go
	modified:   google-beta/services/resourcemanager/data_source_google_service_account_access_token_test.go
	modified:   google-beta/services/resourcemanager/data_source_google_service_account_id_token.go
	modified:   google-beta/services/resourcemanager/data_source_google_service_account_id_token_test.go
	modified:   google-beta/services/resourcemanager/data_source_google_service_account_jwt_test.go
	modified:   google-beta/services/resourcemanager/resource_google_project_migrate_test.go
	modified:   google-beta/services/resourcemanager/resource_google_service_account_key_test.go
	modified:   google-beta/services/resourcemanager/resource_google_service_account_sweeper.go
```

@rileykarson appreciate your help here.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 76 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 76 insertions(+), 2 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `136` 
Passed tests `105` 
Skipped tests: `30` 
Affected tests: `1` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccServiceAccount_createIgnoreAlreadyExists </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{red}{\textsf{The provider crashed while running the VCR tests in RECORDING mode}}$ 
$\textcolor{red}{\textsf{Please fix it to complete your PR}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/394059b5-71dc-4fdc-baac-5ece3cb8763e/build-log/recording_test.log)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 4 files changed, 78 insertions(+), 7 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 78 insertions(+), 2 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `136` 
Passed tests `105` 
Skipped tests: `30` 
Affected tests: `1` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccServiceAccount_createIgnoreAlreadyExists </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 80 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9727-old..auto-pr-9727) ( 3 files changed, 80 insertions(+), 2 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `136` 
Passed tests `105` 
Skipped tests: `30` 
Affected tests: `1` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>resourcemanager</li></ul></blockquote></details> 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccServiceAccount_createIgnoreAlreadyExists </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccServiceAccount_createIgnoreAlreadyExists`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/cceb50e2-3b0b-4cb1-a955-036735421fca/recording/TestAccServiceAccount_createIgnoreAlreadyExists.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/cceb50e2-3b0b-4cb1-a955-036735421fca/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9727/artifacts/cceb50e2-3b0b-4cb1-a955-036735421fca/recording) for each test

benhxy commented:
Tests are fixed.
